frontend/src/metabase/hoc/UserHasSeen/UserHasSeen.tsx

```diff
@@ -0,0 +1,44 @@
+import { useContext, useEffect } from "react";
+
+import { useUserAcknowledgement } from "metabase/hooks/use-user-acknowledgement";
+
+import { UserHasSeenAllContext } from "./UserHasSeenAllContext";
+
+interface UserHasSeenProps {
+  hasSeenKey: string;
+  children: ({
+    isNew,
+    ack,
+  }: {
+    isNew: boolean;
+    ack: () => void;
+  }) => JSX.Element;
+  overrideFn?: (hasSeen: boolean) => boolean;
+}
+
+export const UserHasSeen = ({
+  hasSeenKey,
+  children,
+  overrideFn = hasSeen => !hasSeen,
+}: UserHasSeenProps) => {
+  const indicatorContext = useContext(UserHasSeenAllContext);
+  const { upsertBadge, removeBadge } = indicatorContext ?? {};
+
+  const [hasSeen, { ack, isLoading }] = useUserAcknowledgement(
+    hasSeenKey,
+    true,
+  );
+
+  const isNew = overrideFn(hasSeen);
+
+  useEffect(() => {
+    if (upsertBadge && removeBadge) {
+      if (!isLoading && isNew) {
```

Comment:
would be nice to name these checks like that explain a bit more as to why you're checking these 4 booleans
```ts
const hasUpdateFns = upsertBadge && removeBadge;
const hasLoadedNew = !isLoading && isNew // pick a better name 😅
const isNewAndUpdatable = hasUpdateFns && hasLoadedNew;
if (isNewAndUpdatable) {
  // remaining code
 ```